### PR TITLE
Add `InlineSpan.visitDirectChildren`

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
I'd like to find out the `fontSize` of a `PlaceholderSpan`, and currently there doesn't seem to be a way to do `TextStyle` ~cascading~ inheritance in the framework:

 `InlineSpan.visitChildren` traverses the entire `InlineSpan` tree using a preorder traversal, and nodes that don't have "content" will be skipped (https://master-api.flutter.dev/flutter/painting/InlineSpan/visitChildren.html): 

> Walks this [InlineSpan](https://master-api.flutter.dev/flutter/painting/InlineSpan-class.html) and any descendants in pre-order and calls visitor for each span that has content.

which makes it impossible to do `TextStyle` inheritance in the framework: 
- `InlineSpan`s with a non-null `TextStyle` but has no content will be skipped
- `visitChildren` doesn't directly expose the hierarchy, it only gives information about the flattened tree.

This doesn't look like a breaking change, most internal customers are extending `WidgetSpan` which has a concrete implementation of the new method.

Alternatively I could create a fake `ui.ParagraphBuilder` and record the `ui.TextStyle` at the top of the stack when `addPlaceholder` is called. But `ui.TextStyle` properties are not exposed to the framework.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/painting/inline_span.dart [hunk 1]
```diff
@@ -227,6 +227,18 @@ abstract class InlineSpan extends DiagnosticableTree {
...
+  ///
+  /// The immediate children are visited in the same order they are added to
+  /// a [ui.ParagraphBuilder] in the [build] method, which is also the logical
+  /// order of the child [InlineSpan]s in the text.
+  ///
+  /// The traversal stops when all immediate children are visited, or when the
+  /// `visitor` callback returns `false` on an immediate child. This method
+  /// itself returns a `bool` indicating whether the visitor callback returned
+  /// `true` on all immediate children.
+  bool visitDirectChildren(InlineSpanVisitor visitor);
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/inline_span.dart:240]: Can you link the docs of visitDirectChildren and visitChildren which a one-sentence summary of what the difference is?

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 2]
```diff
@@ -1144,29 +1141,17 @@ class TextPainter {
...
-        }
-      case TextAlign.end:
-        switch (textDirection) {
-          case TextDirection.rtl:
-            return 0.0;
-          case TextDirection.ltr:
-            return 1.0;
-        }
-    }
+    return switch ((textAlign, textDirection)) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/text_painter.dart:1144]: nice usage of new language features!

File: packages/flutter/lib/src/painting/text_span.dart [hunk 3]
```diff
@@ -275,9 +275,9 @@ class TextSpan extends InlineSpan implements HitTestTarget, MouseTrackerAnnotati
     if (hasStyle) {
       builder.pushStyle(style!.getTextStyle(textScaleFactor: textScaleFactor));
     }
-    if (text != null) {
+    if (text case final String text) {
```
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/painting/text_span.dart:278]: Same below in other places in this PR.

File: packages/flutter/lib/src/painting/text_span.dart [hunk 4]
```diff
@@ -310,21 +310,13 @@ class TextSpan extends InlineSpan implements HitTestTarget, MouseTrackerAnnotati
...
-    if (children != null) {
-      for (final InlineSpan child in children!) {
-        if (!child.visitChildren(visitor)) {
-          return false;
-        }
-      }
-    }
-    return true;
+    return (text == null || visitor(this))
+        && !(children?.any((InlineSpan descendant) => !descendant.visitChildren(visitor)) ?? false);
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/text_span.dart:314]: The old code is way more readable here and it also made the intend clearer.

Comment (goderbauer) [reply packages/flutter/lib/src/painting/text_span.dart:314]: (the first two ifs in the old code could probably be merged, though)

File: packages/flutter/lib/src/painting/text_span.dart [hunk 5]
```diff
@@ -310,21 +310,13 @@ class TextSpan extends InlineSpan implements HitTestTarget, MouseTrackerAnnotati
...
   }
 
+  @override
+  bool visitDirectChildren(InlineSpanVisitor visitor) => !(children?.any((InlineSpan child) => !visitor(child)) ?? false);
+
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/text_span.dart:319]: Same, this doesn't seem to be the most readable way of expressing the logic.

**Review by goderbauer — COMMENTED** [review]
It's too bad that the semantics of InlineSpan.visitChildren is different from most other visitChildren, which do not automatically traverse... Oh well.

File: packages/flutter/lib/src/painting/text_span.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/text_span.dart:278]: I wonder what makes the optional binding form harder to read? Was it the lack of syntax highlighting on github?

or would it be cleaner if it's with a `?` or when the type annotation gets omitted?
```dart
  if (text case final text?) {
```
or 
```dart
  if (text case final nonNullText?) {
```

[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/text_span.dart:278]: oh the keywords are highlighted.

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**